File: wrappers/swift/Sources/PrinterDriver/Printer.swift
import CPrinterDriver
import Foundation

/// One configured device: transport, width and capability profile.
///
/// Owned by the ``PrinterDriver`` that created it and valid for that driver's lifetime,
/// which this object extends by holding it.
public final class Printer: @unchecked Sendable {
  let core: DriverCore
  /// Internal, not private, so the test target can reach the same `pd_printer*` the
  /// scripted-device support functions are keyed on.
  let handle: OpaquePointer
  private let lock = NSLock()
  private var deviceTrampoline: DeviceEventTrampoline?

  /// The stable printer id, either the one supplied at configuration or one derived from
  /// the endpoint.
  public let id: String

  /// The printable width the core will scale rasters to.
  public let widthDots: UInt32

  /// Which ordered fence this printer's profile answers.
  ///
  /// This is the ceiling on what a `done` result can ever rest on:
  /// ``CompletionMechanism/gsParenH`` can prove a fault-free cut,
  /// ``CompletionMechanism/gsR1`` can prove ordering, and
  /// ``CompletionMechanism/none`` can prove only that the transport took the bytes.
  public let completionMechanism: CompletionMechanism

  /// What ``completionMechanism`` is actually worth — docs/compatibility-brief.md §28.
  ///
  /// Not a confidence level, and it changes nothing about what a job reports. It answers
  /// a different question, and answers it *before* anything has been printed: should this
  /// printer's fence be trusted on the strength of a profile alone?
  ///
  /// ``Provenance/documented`` means the manufacturer's own command documentation lists
  /// the mechanism for this model — in the shipped database, Epson and nobody else.
  /// ``Provenance/probed`` means this driver asked the installed hardware over the
  /// installed interface path and it answered, which beats any datasheet and is specific
  /// to that path. ``Provenance/unverified`` means neither, which is what "ESC/POS
  /// compatible" on a spec sheet amounts to — and is the signal that running `pdctl probe`
  /// against the site's hardware is worth the trip.
  public let completionProvenance: Provenance

  /// The command language this printer's profile is driven in.
  ///
  /// Anything but ``CommandLanguage/escPos`` is refused with ``FailureReason/unsupported``
  /// before a byte is written: a Zebra or a Brother is not an ESC/POS printer, and sending
  /// it ESC/POS produces a metre of text rather than a label.
  public let language: CommandLanguage

  init(core: DriverCore, handle: OpaquePointer) {
    self.core = core
    self.handle = handle
    id = String(cString: pd_printer_id(handle))
    widthDots = pd_printer_width_dots(handle)
    completionMechanism = CompletionMechanism(bridging: pd_printer_completion(handle).rawValue)
    completionProvenance = Provenance(
      bridging: pd_printer_completion_provenance(handle).rawValue)
    language = CommandLanguage(bridging: pd_printer_language(handle).rawValue)
  }

  // MARK: - Printing

  /// Submits a job.
  ///
  /// Returns as soon as the core has taken the payload; the printing itself is reported
  /// through the returned job's ``PrintJob/events`` and ``PrintJob/result``.
  ///
  /// - Important: Re-submitting a ``JobOptions/key`` that already has a job **does not
  ///   print**. It returns that job — the same object, in whatever state it is in. This
  ///   is the duplicate-ticket defence, and it survives an app restart. To print the same
  ///   key again on purpose, use ``forceReprint(key:options:)``.
  /// - Throws: ``PrinterDriverError`` when the payload is malformed.
  public func print(_ payload: Payload, options: JobOptions = JobOptions()) throws -> PrintJob {
    let handle = try payload.withABI { payloadPointer in
      try options.withABI { optionsPointer -> OpaquePointer in
        guard let job = pd_print(core.handle, self.handle, payloadPointer, optionsPointer) else {
          throw core.lastError()
        }
        return job
      }
    }
    return core.internJob(handle, mechanism: completionMechanism)
  }

  /// Submits a job and reports it through closures — the callback form of
  /// ``print(_:options:)`` for code that is not in an `async` context.
  ///
  /// - Parameters:
  ///   - onProgress: called for every ``JobEvent``, in order, including the ones already
  ///     recorded before this call.
  ///   - completion: called **exactly once**, with the tri-state terminal result. It fires
  ///     after the last `onProgress` call for that job, and it fires for a job that had
  ///     already finished before you asked.
  /// - Returns: the same handle ``print(_:options:)`` would return, so a caller can still
  ///   read state or attach an `AsyncStream` later.
  ///
  /// Both closures are delivered on the driver's serial delivery queue, not on the main
  /// queue and not on a core worker thread.
  @discardableResult
  public func print(
    _ payload: Payload,
    options: JobOptions = JobOptions(),
    onProgress: @escaping (JobEvent) -> Void = { _ in },
    completion: @escaping (JobResult) -> Void
  ) throws -> PrintJob {
    let job = try print(payload, options: options)
    job.observe(onProgress: onProgress, completion: completion)
    return job
  }

  /// Prints a deliberate duplicate of an already-submitted key.
  ///
  /// Reuses the original payload and prepends the reprint banner and attempt counter, so
  /// the paper says what it is. This is the only way to print a key twice.
  ///
  /// - Parameter options: pass ``ReprintOptions/banner`` `false` only for a receipt
  ///   where the banner is inappropriate; the attempt counter still increments either
  ///   way, so the journal records the duplicate whatever the paper says.
  /// - Throws: ``PrinterDriverError`` when the key is unknown, or when its job was
  ///   reconstructed from the journal — those records carry what happened to a job, never
  ///   what it contained.
  public func forceReprint(key: String, options: ReprintOptions = ReprintOptions())
    throws -> PrintJob
  {
    var effective = options
    effective.job.key = key
    let handle = try key.withCString { keyPointer in
      try effective.withABI { optionsPointer -> OpaquePointer in
        guard
          let job = pd_force_reprint_opts(
            core.handle, self.handle, keyPointer, optionsPointer)
        else {
          throw core.lastError()
        }
        return job
      }
    }
    return core.internJob(handle, mechanism: completionMechanism)
  }

  /// Prints a deliberate duplicate with the plain submission options, banner on.
  public func forceReprint(key: String, options: JobOptions) throws -> PrintJob {
    try forceReprint(key: key, options: ReprintOptions(options))
  }

  // MARK: - Device state

  /// The last known device state. A snapshot, never a live query, so it cannot block
  /// behind a print.
  public func status() -> DeviceStatus {
    DeviceStatus(pd_printer_status(core.handle, handle))
  }

  /// Queues a status round trip behind any active job and waits for the answer.
  ///
  /// - Parameter timeoutMilliseconds: 0 means the ABI default of 2000.
  /// - Important: Blocks the calling thread. Do not call it from the main thread or from
  ///   inside an event handler.
  public func refreshStatus(timeoutMilliseconds: UInt32 = 0) -> DeviceStatus {
    DeviceStatus(pd_printer_refresh_status(core.handle, handle, timeoutMilliseconds))
  }

  /// The per-printer event stream that replaces availability ping-polling.
  ///
  /// Each access returns a fresh stream carrying events from that point on; there is no
  /// replay, because a device event describes a moment rather than a state. Read
  /// ``status()`` for where things stand now.
  public var deviceEvents: AsyncStream<DeviceEvent> {
    lock.lock()
    let trampoline: DeviceEventTrampoline
    if let existing = deviceTrampoline {
      trampoline = existing
      lock.unlock()
    } else {
      trampoline = DeviceEventTrampoline(delivery: core.delivery)
      deviceTrampoline = trampoline
      lock.unlock()
      // Once only: pd.h has no unsubscribe, so a second subscription would double every
      // event for the life of the driver.
      core.retain(trampoline: trampoline)
      pd_subscribe_device(
        core.handle, handle, deviceEventThunk,
        Unmanaged.passUnretained(trampoline).toOpaque())
    }
    return trampoline.makeStream()
  }

  /// Pulses the cash drawer and tells the caller nothing.
  ///
  /// Kept because it is ABI, and because a caller that genuinely does not want to wait is
  /// entitled to say so. Everything new should use ``openDrawer(channel:pulseMilliseconds:)``,
  /// which reports what actually happened.
  public func openCashDrawer() {
    pd_open_cash_drawer(core.handle, handle)
  }

  // MARK: - M14: cash drawer (docs/cash-drawer.md)

  /// What this printer's drawer port is, and what is known about it.
  public var drawerCapabilities: DrawerCapabilities {
    DrawerCapabilities(pd_printer_drawer_capabilities(handle))
  }

  /// Fires the drawer and reports what was observed, never a boolean.
  ///
  /// The sequence: read the switch first (a drawer that is already out is not pulsed
  /// again), take the printer's peripheral lane so the pulse cannot land inside a fenced
  /// job, emit one queued pulse, watch the switch for the profile's verification window,
  /// and hold the manufacturer cooldown before the next one.
  ///
  /// Refused — zero bytes written, ``DrawerState/unknown`` — when this model has no drawer
  /// port, when its kick method is one this engine does not drive, or when the electrical
  /// standard is ``DrawerPortStandard/unknown``. RJ11/RJ12-looking drawer connectors are
  /// not a universal electrical standard, and an unclassified port is never energised.
  ///
  /// - Parameters:
  ///   - channel: 1 for drive 1 (Epson pin 2), 2 for drive 2 (pin 5). Clamped to what the
  ///     profile documents.
  ///   - pulseMilliseconds: 0 uses the profile's own default, 200 ms on the documented
  ///     24 V families.
  /// - Important: Blocks the calling thread until the sequence reaches a verdict.
  @discardableResult
  public func openDrawer(channel: UInt8 = 1, pulseMilliseconds: UInt16 = 0) -> DrawerResult {
    var request = pd_drawer_request(channel: channel, pulse_ms: pulseMilliseconds)
    return DrawerResult(pd_drawer_open(core.handle, handle, &request))
  }

  /// Reads the drawer switch without firing anything.
  ///
  /// Safe on hardware ``openDrawer(channel:pulseMilliseconds:)`` refuses, which is what
  /// makes it the first half of the calibration procedure.
  ///
  /// - Important: Blocks the calling thread.
  public func readDrawerSensor(timeoutMilliseconds: UInt32 = 1500) -> DrawerReading {
    DrawerReading(pd_drawer_read_sensor(core.handle, handle, timeoutMilliseconds))
  }

  /// Records which sense level means "open" for the drawer attached to this printer, and
  /// persists it in the driver's storage directory.
  ///
  /// The procedure is an operator's: ask for the drawer to be closed, read the sensor, ask
  /// for it to be opened, read again, and pass the level observed while it was open.
  ///
  /// - Returns: `true` when it was persisted; `false` when it applies to this process only
  ///   (an in-memory driver, or a storage directory that cannot be written).
  @discardableResult
  public func calibrateDrawerPolarity(highMeansOpen: Bool) -> Bool {
    pd_drawer_calibrate_polarity(core.handle, handle, highMeansOpen ? 1 : 0) == 1
  }

  /// Whether a polarity has been measured for this printer. Until it has, a sensor reading
  /// carries a level and no interpretation.
  public var isDrawerPolarityCalibrated: Bool {
    pd_drawer_polarity_calibrated(core.handle, handle) == 1
  }

  /// Meaningful only while ``isDrawerPolarityCalibrated`` is true.
  public var drawerHighMeansOpen: Bool {
    pd_drawer_high_means_open(core.handle, handle) == 1
  }

  /// Waits until this printer's queue is empty and its active job is terminal.
  ///
  /// - Important: Blocks the calling thread.
  public func drain() {
    pd_printer_drain(core.handle, handle)
  }
}

File: wrappers/dotnet/PrinterDriver/Printer.cs
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using System.Threading.Channels;

namespace PrinterDriver;

/// <summary>
/// One printer, owned by the <see cref="PrinterDriver"/> that attached it.
/// </summary>
/// <remarks>
/// Nothing here is disposable: the driver owns every handle and frees them all in
/// <see cref="PrinterDriver.Dispose"/>.
/// </remarks>
public sealed partial class Printer
{
    private readonly PrinterDriver _driver;
    private readonly nint _handle;
    private readonly object _gate = new();
    private DeviceSubscription? _subscription;

    internal Printer(PrinterDriver driver, nint handle)
    {
        _driver = driver;
        _handle = handle;
        Id = NativeMethods.ReadUtf8(NativeMethods.pd_printer_id(handle));
    }

    /// <summary>The printer's stable identifier.</summary>
    public string Id { get; }

    /// <summary>Print width in dots.</summary>
    public uint WidthDots => NativeMethods.pd_printer_width_dots(_handle);

    /// <summary>
    /// Which ordered fence this printer's profile answers. <see cref="CompletionMechanism.None"/>
    /// means no job on it can ever do better than <see cref="JobResult.Unknown"/>.
    /// </summary>
    public CompletionMechanism Completion =>
        (CompletionMechanism)NativeMethods.pd_printer_completion(_handle);

    /// <summary>
    /// What the fence <see cref="Completion"/> names is actually worth
    /// (docs/compatibility-brief.md §28).
    /// </summary>
    /// <remarks>
    /// Not a confidence level, and it changes nothing about what a job reports. It answers
    /// a different question, before anything has been printed: should this printer's fence
    /// be trusted on the strength of its profile alone? Epson is the only family whose
    /// shipped defaults say <see cref="Provenance.Documented"/>, and
    /// <see cref="Provenance.Probed"/> means this driver asked the installed hardware over
    /// the installed interface path — which is what makes probing a site's machines worth
    /// the trip.
    /// </remarks>
    public Provenance CompletionProvenance =>
        (Provenance)NativeMethods.pd_printer_completion_provenance(_handle);

    /// <summary>
    /// The language this printer's profile is driven in. Anything but
    /// <see cref="CommandLanguage.EscPos"/> is refused with
    /// <see cref="FailureReason.Unsupported"/> before a byte is written.
    /// </summary>
    public CommandLanguage Language => (CommandLanguage)NativeMethods.pd_printer_language(_handle);

    /// <summary>
    /// Last known device state. Never a live query, so it cannot block behind a print.
    /// </summary>
    public DeviceStatus Status =>
        DeviceStatus.FromNative(NativeMethods.pd_printer_status(_driver.Handle, _handle));

    /// <summary>
    /// Queues a status round trip behind any active job and waits for the answer.
    /// </summary>
    /// <param name="timeoutMs">How long to wait; 0 means the ABI's default of 2000 ms.</param>
    /// <returns>The refreshed state.</returns>
    public DeviceStatus RefreshStatus(uint timeoutMs = 0) => DeviceStatus.FromNative(
        NativeMethods.pd_printer_refresh_status(_driver.Handle, _handle, timeoutMs));

    /// <summary>Kicks the cash drawer and tells the caller nothing.</summary>
    /// <remarks>
    /// Kept because it is ABI, and because a caller that genuinely does not want to wait is
    /// entitled to say so. Everything new should use <see cref="OpenDrawer"/>, which reports
    /// what actually happened.
    /// </remarks>
    public void OpenCashDrawer() => NativeMethods.pd_open_cash_drawer(_driver.Handle, _handle);

    // --- M14: cash drawer (docs/cash-drawer.md) --------------------------------------

    /// <summary>What this printer's drawer port is, and what is known about it.</summary>
    public DrawerCapabilities DrawerCapabilities =>
        global::PrinterDriver.DrawerCapabilities.FromNative(
            NativeMethods.pd_printer_drawer_capabilities(_handle));

    /// <summary>
    /// Fires the drawer and reports what was observed, never a boolean.
    /// </summary>
    /// <remarks>
    /// The sequence: read the switch first (a drawer that is already out is not pulsed
    /// again), take the printer's peripheral lane so the pulse cannot land inside a fenced
    /// job, emit one queued pulse, watch the switch for the profile's verification window,
    /// and hold the manufacturer cooldown before the next one.
    /// <para>
    /// Refused — zero bytes written, <see cref="DrawerState.Unknown"/> — when this model has
    /// no drawer port, when its kick method is one this engine does not drive, or when the
    /// electrical standard is <see cref="DrawerPortStandard.Unknown"/>. RJ11/RJ12-looking
    /// drawer connectors are not a universal electrical standard, and an unclassified port
    /// is never energised.
    /// </para>
    /// <para>Blocks the calling thread until the sequence reaches a verdict.</para>
    /// </remarks>
    /// <param name="channel">
    /// 1 for drive 1 (Epson pin 2), 2 for drive 2 (pin 5). Clamped to what the profile
    /// documents.
    /// </param>
    /// <param name="pulseMs">0 uses the profile's own default.</param>
    /// <returns>What was established.</returns>
    public DrawerResult OpenDrawer(byte channel = 1, ushort pulseMs = 0)
    {
        var request = new PdDrawerRequest { Channel = channel, PulseMs = pulseMs };
        return DrawerResult.FromNative(
            NativeMethods.pd_drawer_open(_driver.Handle, _handle, in request));
    }

    /// <summary>
    /// Reads the drawer switch without firing anything. Safe on hardware
    /// <see cref="OpenDrawer"/> refuses, which is what makes it the first half of the
    /// calibration procedure.
    /// </summary>
    /// <param name="timeoutMs">How long to wait for the answer; 0 means 1500 ms.</param>
    /// <returns>The raw level, and an interpretation only where one has been earned.</returns>
    public DrawerReading ReadDrawerSensor(uint timeoutMs = 0) => DrawerReading.FromNative(
        NativeMethods.pd_drawer_read_sensor(_driver.Handle, _handle, timeoutMs));

    /// <summary>
    /// Records which sense level means "open" for the drawer attached to this printer, and
    /// persists it in the driver's storage directory.
    /// </summary>
    /// <remarks>
    /// The procedure is an operator's: ask for the drawer to be closed, read the sensor, ask
    /// for it to be opened, read again, and pass the level observed while it was open.
    /// </remarks>
    /// <param name="highMeansOpen">The level observed while the drawer was open.</param>
    /// <returns>
    /// True when it was persisted; false when it applies to this process only — an in-memory
    /// driver, or a storage directory that cannot be written.
    /// </returns>
    public bool CalibrateDrawerPolarity(bool highMeansOpen) =>
        NativeMethods.pd_drawer_calibrate_polarity(_driver.Handle, _handle,
                                                    highMeansOpen ? 1 : 0) == 1;

    /// <summary>
    /// Whether a polarity has been measured for this printer. Until it has, a sensor reading
    /// carries a level and no interpretation.
    /// </summary>
    public bool IsDrawerPolarityCalibrated =>
        NativeMethods.pd_drawer_polarity_calibrated(_driver.Handle, _handle) == 1;

    /// <summary>Meaningful only while <see cref="IsDrawerPolarityCalibrated"/> is true.</summary>
    public bool DrawerHighMeansOpen =>
        NativeMethods.pd_drawer_high_means_open(_driver.Handle, _handle) == 1;

    /// <summary>Blocks until this printer's queue is empty and its active job is terminal.</summary>
    public void Drain() => NativeMethods.pd_printer_drain(_driver.Handle, _handle);

    /// <summary>
    /// Submits a job.
    /// </summary>
    /// <remarks>
    /// Re-submitting an idempotency key that already has a job does not print: it returns
    /// that job, whatever state it is in (docs/api.md §3). Because the driver interns
    /// handles, that shows up as the identical <see cref="PrintJob"/> instance.
    /// </remarks>
    /// <param name="payload">What to print.</param>
    /// <param name="options">Key, cut, preflight and timeout; null takes the defaults.</param>
    /// <returns>The job.</returns>
    /// <exception cref="PrinterDriverException">The payload was malformed.</exception>
    public PrintJob Print(Payload payload, JobOptions? options = null)
    {
        ArgumentNullException.ThrowIfNull(payload);
        options ??= new JobOptions();
        using var marshalled = new PayloadMarshaller(payload);
        var key = Marshal.StringToCoTaskMemUTF8(options.Key ?? string.Empty);
        try
        {
            var native = NativeOptions(options, key);
            var payloadStruct = marshalled.Native;
            var job = NativeMethods.pd_print(_driver.Handle, _handle, in payloadStruct, in native);
            if (job == 0)
            {
                throw new PrinterDriverException(_driver.LastError());
            }
            return _driver.InternJob(job);
        }
        finally
        {
            Marshal.FreeCoTaskMem(key);
        }
    }

    /// <summary>
    /// A deliberate duplicate of an already-submitted key: reuses the original payload and
    /// prepends the reprint banner and attempt counter.
    /// </summary>
    /// <param name="key">The original idempotency key.</param>
    /// <param name="options">Options for the reprint; null takes the defaults.</param>
    /// <returns>The new attempt.</returns>
    /// <exception cref="PrinterDriverException">
    /// The key is unknown, or its job was reconstructed from the journal — those records
    /// carry what happened to a job, never what it contained.
    /// </exception>
    public PrintJob ForceReprint(string key, JobOptions? options = null)
    {
        ArgumentException.ThrowIfNullOrEmpty(key);
        options ??= new JobOptions();
        var keyPointer = Marshal.StringToCoTaskMemUTF8(options.Key ?? string.Empty);
        try
        {
            var native = NativeOptions(options, keyPointer);
            var job = NativeMethods.pd_force_reprint(_driver.Handle, _handle, key, in native);
            if (job == 0)
            {
                throw new PrinterDriverException(_driver.LastError());
            }
            return _driver.InternJob(job);
        }
        finally
        {
            Marshal.FreeCoTaskMem(keyPointer);
        }
    }

    /// <summary>
    /// The same reprint, with control over the banner (<c>pd_force_reprint_opts</c>).
    /// </summary>
    /// <remarks>
    /// Suppressing the banner is a per-call, deliberate act for a receipt where it is
    /// inappropriate. A kitchen ticket should never suppress it: the banner is what lets
    /// staff bin the duplicate instead of cooking it twice.
    /// </remarks>
    /// <param name="key">The original idempotency key.</param>
    /// <param name="printBanner">False prints the duplicate with no banner.</param>
    /// <param name="options">Options for the reprint; null takes the defaults.</param>
    /// <returns>The new attempt.</returns>
    /// <exception cref="PrinterDriverException">
    /// The key is unknown, or its job was reconstructed from the journal.
    /// </exception>
    public PrintJob ForceReprint(string key, bool printBanner, JobOptions? options = null)
    {
        ArgumentException.ThrowIfNullOrEmpty(key);
        options ??= new JobOptions();
        var keyPointer = Marshal.StringToCoTaskMemUTF8(options.Key ?? string.Empty);
        try
        {
            var native = new PdReprintOptions
            {
                Job = NativeOptions(options, keyPointer),
                SuppressBanner = printBanner ? 0 : 1,
            };
            var job = NativeMethods.pd_force_reprint_opts(
                _driver.Handle, _handle, key, in native);
            if (job == 0)
            {
                throw new PrinterDriverException(_driver.LastError());
            }
            return _driver.InternJob(job);
        }
        finally
        {
            Marshal.FreeCoTaskMem(keyPointer);
        }
    }

    /// <summary>
    /// Submits a job and awaits its terminal answer.
    /// </summary>
    /// <param name="payload">What to print.</param>
    /// <param name="options">Key, cut, preflight and timeout; null takes the defaults.</param>
    /// <param name="cancellationToken">Stops waiting; the job itself keeps going.</param>
    /// <returns>The tri-state result.</returns>
    public Task<JobResult> SendAsync(Payload payload, JobOptions? options = null,
                                     CancellationToken cancellationToken = default) =>
        Print(payload, options).GetResultAsync(cancellationToken);

    /// <summary>
    /// The device event stream, which is what replaces polling for availability.
    /// </summary>
    /// <remarks>
    /// Every subscriber reads the same underlying native subscription: pd.h has no
    /// unsubscribe, so this subscribes once, on first use, and fans out from there. A
    /// consumer that stops enumerating stops reading; it does not stop the core.
    /// </remarks>
    /// <param name="cancellationToken">Stops the enumeration.</param>
    /// <returns>Device events as they are decoded.</returns>
    public IAsyncEnumerable<DeviceEvent> EventsAsync(
        CancellationToken cancellationToken = default)
    {
        // Deliberately NOT an `async IAsyncEnumerable` method. The body of one of those
        // does not run until the first MoveNextAsync, which would leave a window between
        // "the caller asked for events" and "this printer is actually listening" -- and
        // every event decoded inside that window would be lost. Registering eagerly and
        // returning the reader's own enumerable closes it.
        var reader = Subscription().Add();
        return reader.ReadAllAsync(cancellationToken);
    }

    /// <summary>
    /// Device events as a classic .NET event, for call sites that would rather not
    /// enumerate. Handlers run on whichever core thread decoded the status — the
    /// transport's reader thread, or a worker thread when a status query answers inside a
    /// job — so they must not block and must not call back into this SDK.
    /// </summary>
    public event EventHandler<DeviceEvent>? DeviceEventReceived
    {
        add => Subscription().Handlers += value;
        remove => Subscription().Handlers -= value;
    }

    internal nint Handle => _handle;

    private DeviceSubscription Subscription()
    {
        lock (_gate)
        {
            if (_subscription is not null)
            {
                return _subscription;
            }
            var subscription = new DeviceSubscription(this);
            // Root the state object and the delegate before handing either to the ABI:
            // after pd_subscribe_device returns, a worker thread may call the delegate at
            // any moment, and nothing else in the managed heap refers to it.
            var context = _driver.Roots.Root(subscription);
            var callback = new NativeMethods.DeviceEventCallback(DeviceSubscription.Trampoline);
            _driver.Roots.Root(callback);
            NativeMethods.pd_subscribe_device(_driver.Handle, _handle, callback,
                                              GCHandle.ToIntPtr(context));
            _subscription = subscription;
            return subscription;
        }
    }

    private static PdJobOptions NativeOptions(JobOptions options, nint key) => new()
    {
        Key = key,
        Cut = (int)options.Cut,
        OpenDrawer = options.OpenDrawer ? 1 : 0,
        Preflight = (int)options.Preflight,
        TimeoutMs = options.TimeoutMs,
    };

    /// <summary>The one native device subscription, fanned out to any number of readers.</summary>
    private sealed class DeviceSubscription(Printer printer)
    {
        private readonly List<Channel<DeviceEvent>> _channels = [];
        private readonly object _gate = new();

        internal EventHandler<DeviceEvent>? Handlers;

        internal ChannelReader<DeviceEvent> Add()
        {
            var channel = Channel.CreateUnbounded<DeviceEvent>(new UnboundedChannelOptions
            {
                SingleWriter = false,
                SingleReader = true,
            });
            lock (_gate)
            {
                _channels.Add(channel);
            }
            return channel.Reader;
        }

        internal static void Trampoline(nint _, int deviceEvent, nint context)
        {
            var subscription = CallbackRoots.FromContext<DeviceSubscription>(context);
            subscription?.Publish((DeviceEvent)deviceEvent);
        }

        private void Publish(DeviceEvent deviceEvent)
        {
            Channel<DeviceEvent>[] channels;
            lock (_gate)
            {
                channels = [.. _channels];
            }
            foreach (var channel in channels)
            {
                channel.Writer.TryWrite(deviceEvent);
            }
            // pd.h forbids blocking or re-entering the ABI from a callback, so a handler
            // that throws must not take the worker thread down with it -- and must not be
            // allowed to stop the other handlers either.
            foreach (var handler in Handlers?.GetInvocationList() ?? [])
            {
                try
                {
                    ((EventHandler<DeviceEvent>)handler)(printer, deviceEvent);
                }
                catch (Exception)
                {
                    // Deliberately swallowed: see above.
                }
            }
        }
    }
}
